### ISSUE TITLE
"@Listeners" defined listeners applies to whole suite even if repeated.

### ISSUE THREAD

**Issue by Elisedlund** [cbeust/testng#795]
I have a suite looking like this: 

```
<suite name="Suite">
    <test name="Test">
        <classes>
            <class name="TestClassWithListenerAnnotation"/>
        </classes>
    </test> 
    <test name="Test2">
        <classes>
            <class name="TestClassWithListenerAnnotation"/>
        </classes>
    </test> 
</suite>
```

my problem is that the listeners defined in my testclass with @Listeners applies twice which means that I have 2 identical listeners in the suite: 

(onStart ect actually execute 4 time instead of two because of another issue https://github.com/cbeust/testng/issues/767)

```
onExecutionStart()
onExecutionStart()
[TestNG] Running:
  C:\Users\eeliedl\Desktop\workspaces\workspace2\TestTestNg\src\main\java\testng3.xml

onStart(ISuite suite)
onStart(ISuite suite)
onStart(ISuite suite)
onStart(ISuite suite)
onStart(ITestContext context)
intercept(List<IMethodInstance> methods, ITestContext context)
intercept(List<IMethodInstance> methods, ITestContext context)
onTestStart(ITestResult result)
beforeInvocation(IInvokedMethod method, ITestResult testResult)
beforeInvocation(IInvokedMethod method, ITestResult testResult)
beforeInvocation(IInvokedMethod method, ITestResult testResult)
beforeInvocation(IInvokedMethod method, ITestResult testResult)
afterInvocation(IInvokedMethod method, ITestResult testResult)
afterInvocation(IInvokedMethod method, ITestResult testResult)
afterInvocation(IInvokedMethod method, ITestResult testResult)
afterInvocation(IInvokedMethod method, ITestResult testResult)
onTestSuccess(ITestResult result) 
onFinish(ITestContext context)
onStart(ITestContext context)
intercept(List<IMethodInstance> methods, ITestContext context)
intercept(List<IMethodInstance> methods, ITestContext context)
onTestStart(ITestResult result)
beforeInvocation(IInvokedMethod method, ITestResult testResult)
beforeInvocation(IInvokedMethod method, ITestResult testResult)
beforeInvocation(IInvokedMethod method, ITestResult testResult)
beforeInvocation(IInvokedMethod method, ITestResult testResult)
afterInvocation(IInvokedMethod method, ITestResult testResult)
afterInvocation(IInvokedMethod method, ITestResult testResult)
afterInvocation(IInvokedMethod method, ITestResult testResult)
afterInvocation(IInvokedMethod method, ITestResult testResult)
onTestSuccess(ITestResult result) 
onFinish(ITestContext context)
onFinish(ISuite suite)
onFinish(ISuite suite)
onFinish(ISuite suite)
onFinish(ISuite suite)

===============================================
Suite
Total tests run: 2, Failures: 0, Skips: 0
===============================================

onExecutionFinish()
onExecutionFinish()
```

**Comment by Elisedlund** [comment]
a workaround listener delegator class:
´´´
public final class Listener implements IAnnotationTransformer, IInvokedMethodListener, ITestListener,
        ISuiteListener, IExecutionListener, IMethodInterceptor {

    private static boolean wasCreatedOnceBefore = false;
    private boolean isSecondInstance = false;
    private boolean isOnStartSecondTime = false;
    private boolean isOnFinishSecondTime = false;
    private Set<Integer> invokedMethodHashes = new HashSet<Integer>();

    public Listener() {
        if (wasCreatedOnceBefore) {
            isSecondInstance = true;
        }
        wasCreatedOnceBefore = true;
    }

    @Override
    public void onStart(ISuite suite) {
        if (isSecondInstance) {
            return;
        }
        if (isOnStartSecondTime) {
            return;
        }
        isOnStartSecondTime = true;
        System.out.println("onStart(ISuite suite)");
    }

    @Override
    public void onFinish(ISuite suite) {
        if (isSecondInstance) {
            return;
        }
        if (isOnFinishSecondTime) {
            return;
        }
        isOnFinishSecondTime = true;
        System.out.println("onFinish(ISuite suite)");
    }

    @Override
    public void beforeInvocation(IInvokedMethod method, ITestResult testResult) {
        if (isSecondInstance) {
            return;
        }
        if (invokedMethodHashes.contains(method.hashCode())) {
            return;
        }
        invokedMethodHashes.add(method.hashCode());
        System.out.println("beforeInvocation(IInvokedMethod method, ITestResult testResult)");

    }

    @Override
    public void afterInvocation(IInvokedMethod method, ITestResult testResult) {
        if (isSecondInstance) {
            return;
        }
        if (!invokedMethodHashes.contains(method.hashCode())) {
            return;
        }
        invokedMethodHashes.remove(method.hashCode());
        System.out.println("afterInvocation(IInvokedMethod method, ITestResult testResult)");

    }

    @Override
    public void onExecutionStart() {
        if (isSecondInstance) {
            return;
        }
        System.out.println("onExecutionStart()");

    }

    @Override
    public void onExecutionFinish() {
        if (isSecondInstance) {
            return;
        }
        System.out.println("onExecutionFinish()");

    }

    @Override
    public void transform(ITestAnnotation annotation, Class testClass, Constructor testConstructor, Method testMethod) {
        if (isSecondInstance) {
            return;
        }
        System.out
                .println("transform(ITestAnnotation annotation, Class testClass, Constructor testConstructor, Method testMethod)");
    }

    @Override
    public void onTestStart(ITestResult result) {
        System.out.println("onTestStart(ITestResult result)");
    }

    @Override
    public void onTestSuccess(ITestResult result) {
        System.out.println("onTestSuccess(ITestResult result) ");

    }

    @Override
    public void onTestFailure(ITestResult result) {
        System.out.println("onTestFailure(ITestResult result)");
    }

    @Override
    public void onTestSkipped(ITestResult result) {
        System.out.println("onTestSkipped(ITestResult result)");

    }

    @Override
    public void onTestFailedButWithinSuccessPercentage(ITestResult result) {
        System.out.println("onTestFailedButWithinSuccessPercentage(ITestResult result)");
    }

    @Override
    public void onStart(ITestContext context) {
        System.out.println("onStart(ITestContext context)");

    }

    @Override
    public void onFinish(ITestContext context) {
        System.out.println("onFinish(ITestContext context)");
    }

    @Override
    public List<IMethodInstance> intercept(List<IMethodInstance> methods, ITestContext context) {
        System.out.println("intercept(List<IMethodInstance> methods, ITestContext context)");
        return methods;
    }
}
´´´

**Comment by juherr** [comment]
#781 may fix the problem, could you test with [the latest snapshot version](https://oss.sonatype.org/content/repositories/snapshots/org/testng/testng/6.9.7-SNAPSHOT/)?

**Comment by Elisedlund** [comment]
I tried with the latest snapshot as you suggested, (made sure it was used by cleaning .m2)
However no change at all it still behaves the same :/

**Comment by juherr** [comment]
@Elisedlund I'm not able to reproduce your issue. Could you have a look on #805 and tell me if you see something wrong?

**Comment by Elisedlund** [comment]
Hi 
With the latest 6.9.7-SNAPSHOT it works a lot better now.
The suitelistener executes only once now as expected. 

However the IExecutionListener still executes twice which I think it shouldn't

'''
onExecutionStart()
onExecutionStart()
onStart(ISuite suite)
onStart(ITestContext context)
intercept(List<IMethodInstance> methods, ITestContext context)
intercept(List<IMethodInstance> methods, ITestContext context)
onTestStart(ITestResult result)
beforeInvocation(IInvokedMethod method, ITestResult testResult)
hello2Test()
afterInvocation(IInvokedMethod method, ITestResult testResult)
onTestSuccess(ITestResult result) 
onTestStart(ITestResult result)
beforeInvocation(IInvokedMethod method, ITestResult testResult)
helloTest()
afterInvocation(IInvokedMethod method, ITestResult testResult)
onTestSuccess(ITestResult result) 
onFinish(ITestContext context)
onStart(ITestContext context)
intercept(List<IMethodInstance> methods, ITestContext context)
intercept(List<IMethodInstance> methods, ITestContext context)
onTestStart(ITestResult result)
beforeInvocation(IInvokedMethod method, ITestResult testResult)
hello2Test()
afterInvocation(IInvokedMethod method, ITestResult testResult)
onTestSuccess(ITestResult result) 
onTestStart(ITestResult result)
beforeInvocation(IInvokedMethod method, ITestResult testResult)
helloTest()
afterInvocation(IInvokedMethod method, ITestResult testResult)
onTestSuccess(ITestResult result) 
onFinish(ITestContext context)
onFinish(ISuite suite)
onExecutionFinish()
onExecutionFinish()
'''

**Comment by juherr** [comment]
Right, I'm agree.

And what do you thkin about `IMethodInterceptor`?

**Comment by Elisedlund** [comment]
Yes I think the IMethodInterceptor should probably just be executed once.
Right now its seems to be executed twice no matter of the setup,(Listener annotation, xml suite defined, multiple Test tags etc)  

Wrote a new issue on that: #824

**Comment by juherr** [comment]
Thank! I will try to fix it.

**Comment by liuhaozzu** [comment]
hello, Juherr,
Have this bug fixed?
//Hao Liu.

**Comment by juherr** [comment]
@liuhaozzu The issue is not closed, so I suppose it is not yet fixed. But I know @krmahadevan fixed some listener issues too. Did you try on your side?

**Comment by liuhaozzu** [comment]
@juherr I have tried TestNG 6.10.
When I implement the `IConfigurationListener2, IInvokedMethodListener, ITestListener`.
I find that 
onStart, onTestStart, onTestSuccess, onFinish will be invoked twice.
please refer to the git hub rep for more details. 
https://github.com/liuhaozzu/testng-bugreproduce-invoke-twice
//Hao Liu

**Comment by krmahadevan** [comment]
The output is a bit confusing. Can you please do the below alterations ?


   - Add a constructor to your listener wherein you are printing some msg
   within the constructor.
   - For all the listener implementations in your listener, add a call to
   *this.toString()* so that we can see the hashCode. The expectation is
   that you should only be seeing one hashCode value.

If you could please update your README.md (or) this bug with that output it
would be great.


Thanks & Regards
Krishnan Mahadevan

"All the desirable things in life are either illegal, expensive, fattening
or in love with someone else!"
My Scribblings @ http://wakened-cognition.blogspot.com/
My Technical Scribbings @ http://rationaleemotions.wordpress.com/

**Comment by liuhaozzu** [comment]
@krmahadevan it seems that this bug has been fixed in TestNG 6.11.
And I havd one more question. I found that in the MANIFEST.MF it says that the TestNG 6.11 was "Created-By: 1.8.0_73 (Oracle Corporation 25.73-b02)". 
However, when I check its class version, I found it's "cafe babe 0000 0033". 
And 33 means that it's compiled by JDK1.7 which is conflicted to the meta info.
what do you think about it?
//Hao Liu

**Comment by krmahadevan** [comment]
I think Created-By: 1.8.0_73 (Oracle Corporation 25.73-b02) basically
indicates that actual java version using which the jar was built.

But since TestNG is still built with its compatibility version being JDK7.
See here
<https://github.com/cbeust/testng/blob/master/kobalt/src/Build.kt#L75-L77>
and here <https://github.com/cbeust/testng/blob/master/build.gradle#L49-L50>


Thanks & Regards
Krishnan Mahadevan

"All the desirable things in life are either illegal, expensive, fattening
or in love with someone else!"
My Scribblings @ http://wakened-cognition.blogspot.com/
My Technical Scribbings @ http://rationaleemotions.wordpress.com/

**Comment by juherr** [comment]
> it seems that this bug has been fixed in TestNG 6.11.
